src/s3/commands/uploadFile.ts

```diff
@@ -26,51 +33,117 @@ export interface FileSizeBytes {
- * The user may need to load more pages to see the uploaded file reflected in the tree.
  */
 export async function uploadFileCommand(
-    node: S3BucketNode | S3FolderNode,
+    s3Client: S3Client,
+    nodeOrDocument: S3BucketNode | S3FolderNode | vscode.Uri | undefined,
     fileSizeBytes: FileSizeBytes = statFile,
+    getBucket = promptUserForBucket,
+    getFile = getFileToUpload,
     window = Window.vscode(),
-    commands = Commands.vscode(),
-    outputChannel = ext.outputChannel
+    outputChannel = ext.outputChannel,
+    commands = Commands.vscode()
 ): Promise<void> {
-    getLogger().debug('UploadFile called for %O', node)
+    let key: string
+    let bucket: S3.Bucket
+    let file: vscode.Uri | undefined
+    let node: S3BucketNode | S3FolderNode | undefined
+    let document: vscode.Uri | undefined
 
-    const fileLocation = await promptForFileLocation(window)
-    if (!fileLocation) {
-        getLogger().info('UploadFile cancelled')
-        telemetry.recordS3UploadObject({ result: 'Cancelled' })
-        return
+    if (nodeOrDocument) {
+        if ((nodeOrDocument as any).getChildren) {
+            node = nodeOrDocument as S3BucketNode | S3FolderNode
+            document = undefined
+        } else {
+            node = undefined
+            document = nodeOrDocument as vscode.Uri
+        }
+    } else {
+        node = undefined
+        document = undefined
     }
 
-    const fileName = path.basename(fileLocation.fsPath)
-    const key = node.path + fileName
-    const destinationPath = readablePath({ bucket: node.bucket, path: key })
-    try {
-        showOutputMessage(`Uploading file from ${fileLocation} to ${destinationPath}`, outputChannel)
+    if (node) {
+        file = await getFile(undefined, window)
+        if (!file) {
+            showOutputMessage('No file selected, cancelling upload', outputChannel)
+            getLogger().info('UploadFile cancelled')
+            telemetry.recordS3UploadObject({ result: 'Cancelled' })
+            return
+        }
+        key = node.path + path.basename(file.fsPath)
+        bucket = { Name: node.bucket.name }
+    } else {
+        while (true) {
+            file = await getFile(document, window)
+            if (file) {
+                let bucketResponse: S3.Bucket | string
+                try {
+                    bucketResponse = await getBucket(s3Client)
+                } catch (e) {
+                    telemetry.recordS3UploadObject({ result: 'Failed' })
+                    getLogger().error('getBucket failed', e)
+                    return
+                }
 
-        await uploadWithProgress({ node, key, fileLocation, fileSizeBytes: fileSizeBytes(fileLocation), window })
+                if (bucketResponse === 'back') {
+                    continue
+                }
+                if (bucketResponse == 'cancel') {
+                    showOutputMessage('No bucket selected, cancelling upload', outputChannel)
```

Comment:
Should I then?

Reply:
Yes please! Anything shown in a user-interactible way should be localized; the only things you shouldn't localize are anything in a call from a `getLogger()` since we need to make sure we can read those.